Core/include/Acts/Utilities/TypeDispatcher.hpp

```diff
@@ -0,0 +1,189 @@
+// This file is part of the ACTS project.
+//
+// Copyright (C) 2016 CERN for the benefit of the ACTS project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include <format>
+#include <functional>
+#include <type_traits>
+#include <typeindex>
+#include <typeinfo>
+#include <unordered_map>
+#include <vector>
+
+#include <boost/core/demangle.hpp>
+
+namespace Acts {
+
+/// Template class for type-based function dispatch
+///
+/// This class allows registering function pointers associated with specific
+/// derived types of a base class. When invoked with a base class reference,
+/// it will look up and call the appropriate registered function.
+///
+/// @tparam base_t The base class type that will be the first parameter
+/// @tparam signature_t The function signature (e.g., return_t(Args...))
+template <typename base_t, typename signature_t>
+class TypeDispatcher;
+
+template <typename base_t, typename return_t, typename... args_t>
+class TypeDispatcher<base_t, return_t(args_t...)> {
+ public:
+  // Type aliases for frequently used template parameters
+  using base_type = base_t;
+  using return_type = return_t;
+  using self_type = TypeDispatcher<base_type, return_type(args_t...)>;
+
+  using function_signature = return_t(const base_t&, args_t...);
+  using function_type = std::function<function_signature>;
+
+  /// Default constructor
+  TypeDispatcher() = default;
+
+  /// Constructor that registers multiple function pointers with auto-detected types
+  /// @param funcs Function pointers - derived types are auto-detected from their first parameters
```

Comment:
I don't understand the comment `from their first parameters` here, can you elaborate?

Reply:
So basically, you have a function like

```cpp
class Base {};
class Derived : public Base{} ;

void process(const Derived& derived, int a, double b);
```

which you want the thing to call if you have an instance of `Derived`.

You can register it explicitly, like

```cpp
dispatcher.registerFunction<Derived>(process)
```

but you don't have to, the deduction here will figure out `Derived`

```cpp
dispatcher.registerFunction(derived)
```

This comment here refers to the constructor, which does the same thing. You can just say:

```cpp
TypeDispatcher<Base, void(int, double)> dispatcher{
  process // The constructor will deduct that this is a function pointer meant for `Derived`.
};
```

Does that clear it up?